File: trunk/java/UMLGraph/src/org/daisy/urakawa/media/data/utilities/CollectManagedMediaTreeNodeVisitor.java
package org.daisy.urakawa.media.data.utilities;

import java.util.LinkedList;
import java.util.List;

import org.daisy.urakawa.core.TreeNode;
import org.daisy.urakawa.core.visitor.TreeNodeVisitor;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.media.data.ManagedMedia;
import org.daisy.urakawa.property.Property;
import org.daisy.urakawa.property.channel.Channel;
import org.daisy.urakawa.property.channel.ChannelDoesNotExistException;
import org.daisy.urakawa.property.channel.ChannelsProperty;

/**
 * Visitor that collects all MediaData used by the visited TreeNodes
 */
public class CollectManagedMediaTreeNodeVisitor implements TreeNodeVisitor {
	private List<ManagedMedia> mCollectedMedia = new LinkedList<ManagedMedia>();

	/**
	 * @return
	 */
	public List<ManagedMedia> getListOfCollectedMedia() {
		return mCollectedMedia;
	}

	public void preVisit(TreeNode node) {
		// TODO: Why do we need a cast here to compile ?? (List<Property>)  
		for (Property prop : (List<Property>) node.getListOfProperties()) {
			if (prop instanceof ChannelsProperty) {
				ChannelsProperty chProp = (ChannelsProperty) prop;
				for (Channel ch : chProp.getListOfUsedChannels()) {
					try {
						if (chProp.getMedia(ch) instanceof ManagedMedia) {
							ManagedMedia mm = (ManagedMedia) chProp
									.getMedia(ch);
							if (!mCollectedMedia.contains(mm))
								mCollectedMedia.add(mm);
						}
					} catch (MethodParameterIsNullException e) {
						// Should never happen
						throw new RuntimeException("WTF ??!");
					} catch (ChannelDoesNotExistException e) {
						// Should never happen
						throw new RuntimeException("WTF ??!");
					}
				}
			}
		}
	}

	public void postVisit(@SuppressWarnings("unused")
	TreeNode node) {
		return;
	}
}

File: csharp/core/media/data/utilities/CollectManagedMediaTreeNodeVisitor.cs
using System;
using System.Collections.Generic;
using System.Text;
using urakawa.core;
using urakawa.core.visitor;
using urakawa.property.channel;

namespace urakawa.media.data.utilities
{
    /// <summary>
    /// Visitor that collects all <see cref="MediaData"/> used by the visited <see cref="TreeNode"/>s.
    /// </summary>
    public class CollectManagedMediaTreeNodeVisitor : ITreeNodeVisitor
    {
        private List<IManaged> mCollectedMedia = new List<IManaged>();

        /// <summary>
        /// Gets the list of collected <see cref="IManaged"/>
        /// </summary>
        /// <returns>The list</returns>
        /// <remarks>
        /// The returned list is a reference to the list in the <see cref="CollectManagedMediaTreeNodeVisitor"/> instance, 
        /// any changes made to the returned list will reflect in 
        /// the <see cref="CollectManagedMediaTreeNodeVisitor"/> instance
        /// </remarks>
        public List<IManaged> ListOfCollectedMedia
        {
            get { return mCollectedMedia; }
        }

        #region ITreeNodeVisitor Members

        /// <summary>
        /// Any <see cref="IManaged"/> used by the 
        /// </summary>
        /// <param name="node">The node being visited</param>
        /// <returns><c>true</c></returns>
        public bool PreVisit(TreeNode node)
        {
            foreach (urakawa.property.Property prop in node.GetListOfProperties())
            {
                if (prop is ChannelsProperty)
                {
                    ChannelsProperty chProp = (ChannelsProperty) prop;
                    foreach (Channel ch in chProp.ListOfUsedChannels)
                    {
                        if (chProp.GetMedia(ch) is IManaged)
                        {
                            IManaged mm = (IManaged) chProp.GetMedia(ch);
                            if (!mCollectedMedia.Contains(mm)) mCollectedMedia.Add(mm);
                        }
                    }
                }
            }
            return true;
        }

        /// <summary>
        /// Nothing is done in post visit
        /// </summary>
        /// <param name="node">The node being visited</param>
        public void PostVisit(TreeNode node)
        {
            return;
        }

        #endregion
    }
}